src/Data/Functor/Linear/Internal/Applicative.hs

```diff
@@ -81,10 +124,106 @@ instance (Applicative f, Semigroup a) => Semigroup (Ap f a) where
+  Applicative (Generically1 f)
+  where
+  pure = Generically1 Prelude.. genericPure
+  liftA2 f (Generically1 x) (Generically1 y) = Generically1 (genericLiftA2 f x y)
+
+genericPure ::
+  forall f a.
+  (Generic1 f, GApplicative ('ShowType f) (Rep1 f)) =>
+  a ->
+  f a
+genericPure = to1 Prelude.. gpure @('ShowType f)
+
+genericLiftA2 ::
+  forall f a b c.
+  (Generic1 f, GApplicative ('ShowType f) (Rep1 f)) =>
+  (a %1 -> b %1 -> c) ->
+  f a %1 ->
+  f b %1 ->
+  f c
+genericLiftA2 f x y = to1 (gliftA2 @('ShowType f) f (from1 x) (from1 y))
+
+class GApplicative (s :: ErrorMessage) f where
+  gpure :: a -> f a
+  gliftA2 :: (a %1 -> b %1 -> c) -> f a %1 -> f b %1 -> f c
+
+instance
+  Unsatisfiable
+    ( 'Text "Cannot derive a data Applicative instance for" ':$$: s
+        ':$$: 'Text "because empty types cannot implement pure."
+    ) =>
+  GApplicative s V1
+  where
+  gpure = unsatisfiable
+  gliftA2 = unsatisfiable
+
+instance GApplicative s U1 where
+  gpure _ = U1
+  gliftA2 _ U1 U1 = U1
+  {-# INLINE gpure #-}
+  {-# INLINE gliftA2 #-}
+
+instance GApplicative s f => GApplicative s (M1 i c f) where
+  gpure = M1 Prelude.. gpure @s
+  gliftA2 f (M1 x) (M1 y) = M1 (gliftA2 @s f x y)
+  {-# INLINE gpure #-}
+  {-# INLINE gliftA2 #-}
+
+instance GApplicative s Par1 where
+  gpure = Par1
+  gliftA2 f (Par1 x) (Par1 y) = Par1 (f x y)
+  {-# INLINE gpure #-}
+  {-# INLINE gliftA2 #-}
+
+instance (GApplicative s f, Applicative g) => GApplicative s (f :.: g) where
+  gpure = Comp1 Prelude.. gpure @s Prelude.. pure
+  gliftA2 f (Comp1 x) (Comp1 y) = Comp1 (gliftA2 @s (liftA2 f) x y)
+  {-# INLINE gpure #-}
+  {-# INLINE gliftA2 #-}
+
+instance (GApplicative s f, GApplicative s g) => GApplicative s (f :*: g) where
+  gpure a = gpure @s a :*: gpure @s a
+  gliftA2 f (a1 :*: a2) (b1 :*: b2) = gliftA2 @s f a1 b1 :*: gliftA2 @s f a2 b2
+  {-# INLINE gpure #-}
+  {-# INLINE gliftA2 #-}
+
+instance
+  Unsatisfiable
+    ( 'Text "Cannot derive a data Applicative instance for" ':$$: s
+        ':$$: 'Text "because sum types do not admit a uniform Applicative definition."
+    ) =>
+  GApplicative s (x :+: y)
+  where
+  gpure = unsatisfiable
+  gliftA2 = unsatisfiable
+
+instance GApplicative s f => GApplicative s (MP1 m f) where
+  gpure a = MP1 (gpure @s a)
+  gliftA2 f (MP1 a) (MP1 b) = MP1 (gliftA2 @s f a b)
+  {-# INLINE gpure #-}
+  {-# INLINE gliftA2 #-}
+
+instance Monoid c => GApplicative s (K1 i c) where
+  gpure _ = K1 mempty
+  gliftA2 _ (K1 x) (K1 y) = K1 (x <> y)
+  {-# INLINE gpure #-}
+  {-# INLINE gliftA2 #-}
 
-instance Monoid a => Applicative ((,) a) where
-  pure x = (mempty, x)
-  (u, f) <*> (v, x) = (u <> v, f x)
+instance
+  Unsatisfiable
+    ( 'Text "Cannot derive a data Applicative instance for" ':$$: s
+        ':$$: 'Text "because it contains one or more primitive unboxed fields."
+        ':$$: 'Text "Such unboxed types lack canonical monoid operations."
```

Comment:
I guess that there could be a notion of monoid parametric in the type representation. But this is definitely not a battle for today.

Reply:
That wouldn't help without some kind of overhaul of the (currently very limited) mechanism for generic representations of unlifted types.